doc/developer/design/20220314_linearized_reads.md

```diff
@@ -0,0 +1,160 @@
+- What syntax is best to express linearizable (or non) reads? For example, as part of `BEGIN`?
+- Should be eventually change the default experience of using `materialized` to be `CONSISTENT`?
+
+## Formalism and constraints
+
+**Note**
+Please refer to [formalism](https://github.com/MaterializeInc/materialize/blob/fe0f4bac40db1afd21606428be14197601bb5a1c/doc/developer/platform/formalism.md)
+for more background about some of the terms used here, as well as
+this [this jepsen map on consistency](https://jepsen.io/consistency) for details
+about specific consistency models.
+
+Materialized offers reads with stronger _real-time_ consistency guarantees than the default.
+
+Reads again some view `V` always happen at a singular `timestamp`, which must be bound
+by the `upper` and `since` of that view. That view `V` is composed of many such _source_
+pTVC's, which are pTVC's that correspond to the data produced some some _external system_.
+
+True _linearizability_ requires that reads and writes happen in some _total order_ that is
+consistent with the _real time ordering_ of those operations. That is, each operation *appears*
+to happen atomically at some singular point in time bounded by the beginning and end of that
+operation.
+
+A `LINEARIZABLE` read against some view `V` is possible if-and-only-if the following are true:
+
+* All _source_ pTVC's that are at the base of `V` offer a _linearizable_ API for obtaining the
+  latest offset of data for that pTVC.
+  * "offset" here refers to some monotonically increasing identifier that durably and consistently
+  labels data.
+  * A given API is _linearizable_ if-and-only-if later (as in, later in _real-time_) calls tho
+  that API can never produce lower "offets".
+* For all _source_ pTVC's involved, materialized durably associate _real-time_ timestamps
+with specific offsets (these are typically referred to as _timestamp bindings_ and are currently
+implemented for Kafka sources).
+* ...and the responses from those API's can be given some _real-time total order_.
+  * The simplest way to obtain this constraint is to involve only one source.
+
+With the above constraints, the mechanism that materialized uses to service the read
+against some view `V` is as follows:
+
+1. Fetch the latest offsets for all _source_ pTVC's involved.
+2. Wait for timestamps to be durably recorded associated with all offsets, as `source_timestamps`
+3. Choose a timestamp that is _at least_
+`max(max(source_uppers), max(source_timestamps))`. Note that this may require one of both of
+holding back the `since` of `V` or waiting for the `upper` to progress past the chosen timestamp.
+4. Service the read for the view at that timestamp. This will durably advance the `upper`s of all involved
+_source_ pTVC's to at least that timestamp.
+
+This provides _linearizability_ because:
+* materialized guarantees (possible with a timestamp oracle service) that all calls to
+`current_real_time()` monotonically increase.
```

Comment:
This is the first reference to this function, so I'm not sure where you think it's being used elsewhere or how it contributes overall to linearizability.

Reply:
oh this is old, I need to remove this!